### PR TITLE
Add explicit adjoint for `SwapWithZero` and classical simulation tests for decomposition of `SelectSwapQROM`

### PR DESCRIPTION
**Pull request by tanujkhattar**
Part of a series of PRs to address #368 

This PR adds:

- Explicit on_classical_vals and adjoint for `SwapWithZero` bloq. 
- Lots of tests to verify classical simulation of `SelectSwapQROM` (and it's decomposition) and `SwapWithZero`.

### CODE REVIEW

File: qualtran/bloqs/swap_network/swap_with_zero.py [hunk 1]
```diff
@@ -101,54 +136,28 @@ def signature(self) -> Signature:
...
+        selection_bitsizes = cast(Tuple[int, ...], self.selection_bitsizes)
+        n_target_registers = cast(Tuple[int, ...], self.n_target_registers)
+        ret = [*_swap_with_zero_swap_sequence(selection_bitsizes, n_target_registers)]
+        return tuple(ret[::-1] if self.uncompute else ret)
 
     def build_composite_bloq(
         self, bb: 'BloqBuilder', targets: NDArray['Soquet'], **sel: 'Soquet'  # type: ignore[type-var]
     ) -> Dict[str, 'SoquetT']:
-        self.build_via_tree(bb, sel, targets, ())
+        # self.build_via_tree(bb, sel, targets, ())
```
Comment (fdmalone) [inline qualtran/bloqs/swap_network/swap_with_zero.py:151]: remove commented out code.

File: qualtran/bloqs/swap_network/swap_with_zero.py [hunk 2]
```diff
@@ -171,6 +180,19 @@ def wire_symbol(self, reg: Register, idx: Tuple[int, ...] = tuple()) -> 'WireSym
             return TextBox(f'swap{subscript}')
         raise ValueError(f'Unrecognized register name {name}')
 
+    def on_classical_vals(
+        self, *, targets: 'ClassicalValT', **selection: 'ClassicalValT'
+    ) -> Dict[str, 'ClassicalValT']:
+        assert isinstance(targets, np.ndarray)
+        selection_idx = tuple(selection.values())
+        for i, sel_idx_small, idx_one, idx_two in self._swap_sequence:
+            if selection_idx[i] & (1 << sel_idx_small) != 0:
```
Comment (fdmalone) [inline qualtran/bloqs/swap_network/swap_with_zero.py:189]: Add a note to help remind the reader why this bit pattern results in a swap :)

**Review by fdmalone — APPROVED** [review]

File: qualtran/bloqs/swap_network/swap_with_zero.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/bloqs/swap_network/swap_with_zero.py:189]: This is just checking whether the `sel_idx_small`'th bit is 1 or 0 in the selection register. There is a detailed explanation in `_swap_with_zero_swap_sequence` which yields a sequence of swaps to enable the swap with zero.

**Comment by tanujkhattar** [comment]
Thanks, merging now.

File: qualtran/bloqs/swap_network/swap_with_zero.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/bloqs/swap_network/swap_with_zero.py]: done

File: qualtran/bloqs/swap_network/swap_with_zero.py (hunk 2 above)
Comment (fdmalone) [reply qualtran/bloqs/swap_network/swap_with_zero.py:188]: I know, I was hoping there was a TLDR for the _swap_with_zero_sequence

Comment (tanujkhattar) [reply qualtran/bloqs/swap_network/swap_with_zero.py:209]: Added a detailed docstring to `_swap_with_zero_swap_sequence`

**Pull request merged**